README.md

```diff
@@ -3,21 +3,88 @@
+Compared to other blockchain privacy techniques, like zero knowledge proofs, the main appeal of HoneyBadgerMPC is that it is much more flexible. Smart contracts that can perform arbitrary computations on any of the secret data.
+While there are many MPC implementations, HoneyBadgerMPC is uniquely suited for blockchain integration because of its focus on robustness --- it is the first MPC toolkit to provide guaranteed output in spite of Byzantine faults.
 
-HoneyBadgerMPC is the first system implementation of asynchronous MPC.
+## How it Works
+Secure Multiparty Computation (MPC) is about computing on secret-shared data. For each piece of confidential data `x`, each of the `n` server nodes stores a different share `[x]`. Any `t` of the servers can be compromised without revealing any information about the confidential data. However, as long as `n-t` parties are runnng correctly, they can work together to perform arbitrary computations and disclose only the outputs, leaking no additional information about the inputs.
 
-HoneyBadgerMPC is the first MPC toolkit to provide all the following desired
-properties:
+The HoneyBadgerMPC protocol is based on known MPC techniques, carefully selected to achieve the robustness goals. HoneyBadgerMPC consists of three main phases:
 
-* Scales to large number of nodes (5 to 50). *Viff fails at this!*
-* Guarantees output even if some nodes crash or are compromised recent
-  developments like SPDZ and EMP-Toolkit fail at this.
-* Prevents data breaches even if some nodes are compromised.
+1. Collecting client inputs. 
+  Clients submit input to the servers through an input masking technique [COPS15]. The servers start with a secret sharing of a random mask `[r]`. A client retrieves shares of this mask from the servers and reconstructs `r`, and then publishes their masked message `(m+r)`. The servers obtain their share of the input as `[m] := (m+r)-[r]`.
+  
+Our programming model is mainly inspired by Viff. In fact our codebase began as a rewrite of Viff, porting it to `Python3/asyncio` rather than `Python2/Twisted`.
+The programming model is based on Python mixins, which extend a `Share` that represents a pipelined MPC computation.
 
-## [HoneyBadgerMPC Subprotocols](docs/subprotocols.rst)
-See [docs/subprotocols.rst](docs/subprotocols.rst).
+To reach agreement on the published inputs, we make use of either the built-in asynchronous broadcast protocol (a port of HoneyBadgerBFT), or else an external blockchain service (See Fabric Integration and Ethereum Integration).
+
+2. Online phase.
+   Once clients provide input, the protocol computes an MPC program layer by layer. 
+   An MPC program consist of linear operations on secret shared data, as well as batch reconstruction. Linear operations can be computed locally, non-linear operations like multiplication must be emulated using share reconstruction and preprocessed values.
+   See `mpc.py`, `batch_reconstruction.py`, `taskprogramrunner.py`, `reed_solomon.py`
```

Comment:
We should make these actual links to the files

Reply:
Could you take care of this ^